### PR TITLE
Make FFaker::SSN.ssn generate only valid SSNs

### PR DESCRIPTION
**Pull request by lilsweetcaligula**
https://github.com/ffaker/ffaker/issues/348

From [wikipedia](https://en.wikipedia.org/wiki/Social_Security_number#Valid_SSNs):

```
Some special numbers are never allocated:

Numbers with all zeros in any digit group (000-##-####, ###-00-####, ###-##-0000).
Numbers with 666 or 900-999 (Individual Taxpayer Identification Number) in the first digit group.
```

Tests included.

Please let me know if there's something wrong and I will gladly address any changes requests.

Have a nice day!

### CODE REVIEW

File: test/test_ssn.rb [hunk 1]
```diff
@@ -7,7 +7,46 @@ class TestSSN < Test::Unit::TestCase
...
+    numbers = groups
+      .map { |g| strip_leading_zeros(g) }
+      .map { |g| Integer(g) }
+
+    numbers
+  end
+
+  def strip_leading_zeros(s)
+    s.gsub(/\A0+(?!\Z)/, '')
   end
```
Comment (lilsweetcaligula) [inline test/test_ssn.rb]: This helper method strips leading zeros. It handles the case with a string of all zeros by returning a single `0` digit:
```
> strip_leading_zeros('00000')
0
> strip_leading_zeros('0101')
101
```

Please let me know if you find this implementation with the use of regex unreadable and I will gladly provide a non-regex solution.

File: test/test_ssn.rb [hunk 2]
```diff
@@ -7,7 +7,46 @@ class TestSSN < Test::Unit::TestCase
...
+
+  def ssn_to_number_groups(ssn)
+    groups = ssn.split('-')
+
+    numbers = groups
+      .map { |g| strip_leading_zeros(g) }
+      .map { |g| Integer(g) }
+
+    numbers
+  end
```
Comment (lilsweetcaligula) [inline test/test_ssn.rb]: `Kernel::Integer` disallows leading zeros in the numerical representations of integers:
```
> Integer('09')
invalid value for Integer(): "09"
```

For that reason we may need to strip the leading zeros first.

An alternative solution of using `to_i` was considered:
```
> '09'.to_i
9
```

However, `Integer` is more strict and may serve as an extra line of defense if any invalid numeric representation does somehow make it through the tests.

If you find the usage of `#to_i` acceptable, please let me know, and I will push the change.

File: test/test_ssn.rb [hunk 3]
```diff
@@ -7,7 +7,46 @@ class TestSSN < Test::Unit::TestCase
...
+  def test_ssn_second_group_non_zero
+    _first, second_group, _third = ssn_to_number_groups(@actual_ssn)
+
+    assert(second_group.positive?)
+  end
+
+  def test_ssn_third_group_non_zero
+    *_other, third_group = ssn_to_number_groups(@actual_ssn)
+
+    assert(third_group.positive?)
```
Comment (marocchino) [inline test/test_ssn.rb]: `positive?` can not use in older version of ruby
```
/home/travis/build/ffaker/ffaker/test/test_ssn.rb:28:in `test_ssn_second_group_non_zero'
     25:   def test_ssn_second_group_non_zero
     26:     _first, second_group, _third = ssn_to_number_groups(@actual_ssn)
     27: 
  => 28:     assert(second_group.positive?)
     29:   end
     30: 
     31:   def test_ssn_third_group_non_zero
Error: test_ssn_second_group_non_zero(TestSSN): NoMethodError: undefined method `positive?' for 79:Fixnum
```

File: lib/ffaker/ssn.rb [hunk 4]
```diff
@@ -11,7 +11,14 @@ module SSN
     # http://en.wikipedia.org/wiki/Social_Security_number
     #
     def ssn
-      FFaker.numerify('###-##-####')
+      first_group  = fetch_sample([rand(1..665), rand(667..899)])
```
Comment (marocchino) [inline lib/ffaker/ssn.rb]: (maybe) `[*1..665, *667..899].sample`

File: test/test_ssn.rb [hunk 5]
```diff
@@ -7,7 +7,45 @@ class TestSSN < Test::Unit::TestCase
...
+  def setup
+    @actual_ssn = FFaker::SSN.ssn
+  end
+
+  def test_ssn_format
+    assert_match(/\A\d{3}-\d{2}-\d{4}\Z/, @actual_ssn)
+  end
+
+  def test_ssn_first_group_in_valid_range
+    first_group, *_other = ssn_to_number_groups(@actual_ssn)
```
Comment (marocchino) [inline test/test_ssn.rb]: I think it is too much.

```ruby
assert_no_match(/\A666/, @actual_ssn)
assert_match(/\A[0-8]\d{2}-\d{2}-\d{4}\Z/, @actual_ssn)
```

will be simpler.

File: test/test_ssn.rb [hunk 6]
```diff
@@ -7,7 +7,45 @@ class TestSSN < Test::Unit::TestCase
...
+
+    assert(first_group != 666)
+    assert(first_group < 900)
+  end
+
+  def test_ssn_second_group_non_zero
+    _first, second_group, _third = ssn_to_number_groups(@actual_ssn)
+
+    assert(second_group.positive?)
```
Comment (marocchino) [inline test/test_ssn.rb]: same here. 
```rb
assert_no_match(/-00-/, @actual_ssn)
```
will do same thing

File: test/test_ssn.rb [hunk 7]
```diff
@@ -7,7 +7,45 @@ class TestSSN < Test::Unit::TestCase
...
+  def test_ssn_second_group_positive
+    _first, second_group, _third = ssn_to_number_groups(@actual_ssn)
+
+    assert(second_group > 0)
+  end
+
+  def test_ssn_third_group_positive
+    *_other, third_group = ssn_to_number_groups(@actual_ssn)
+
+    assert(third_group > 0)
```
[the remark this answers is not in the document]
Comment (marocchino) [reply test/test_ssn.rb]: 🤦‍♂️

File: test/test_ssn.rb (hunk 2 above)
Comment (marocchino) [reply test/test_ssn.rb]: I think this private method not needed.

```rb
_first, second_group, _third = @actual_ssn.split('-')
assert(second_group != '00')

# or

assert_match(/-00-/, @actual_ssn)
```

will do samething.

**Comment by lilsweetcaligula** [comment]
@marocchino I really did not see the comment, sorry... I will address this right away.

File: test/test_ssn.rb [hunk 8]
```diff
@@ -7,7 +7,21 @@ class TestSSN < Test::Unit::TestCase
...
+  end
+
+  def test_ssn_format
+    assert_no_match(/\A666/, @actual_ssn)
+    assert_match(/\A[0-8]\d{2}-\d{2}-\d{4}\Z/, @actual_ssn)
+  end
+
+  def test_ssn_groups_non_zero
+    first_group, second_group, third_group = @actual_ssn.split('-')
+    with_all_zeros = /\A0+\Z/
```
Comment (lilsweetcaligula) [inline test/test_ssn.rb]: Thanks to your comment, I noticed that my test harness was not testing groups adequately enough.

This test should verify that none of the number groups is an all-zero group, which would invalidate an SSN.

I hope the name for the variable `with_all_zeros`, although not exactly intention-revealing, will read nicely in the code.

**Comment by marocchino** [comment]
Thanks!

**Pull request merged**